### PR TITLE
DM-42159:  eo_pipe/cp_verify parity: bias

### PR DESCRIPTION
**Pull request by czwa**
This adds 
 * bias shift calculation
 * amp correlation calculation

### CODE REVIEW

File: python/lsst/ip/isr/isrStatistics.py [hunk 1]
```diff
@@ -115,6 +117,43 @@ class IsrStatisticsTaskConfig(pexConfig.Config):
...
+        default=1.0/15.0,
+    )
+    biasShiftWindow = pexConfig.Field(
+        dtype=int,
+        doc="Filter window size for highpass filter.",
+        default=30,
+    )
+    biasShiftThreshold = pexConfig.Field(
+        dtype=float,
+        doc="Threshold for bias shift detection.",
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:142]: What units is this?

File: python/lsst/ip/isr/isrStatistics.py [hunk 2]
```diff
@@ -115,6 +117,43 @@ class IsrStatisticsTaskConfig(pexConfig.Config):
...
+        default=5,
+    )
+    biasShiftCutoff = pexConfig.Field(
+        dtype=float,
+        doc="Cutoff frequency for highpass filter.",
+        default=1.0/15.0,
+    )
+    biasShiftWindow = pexConfig.Field(
+        dtype=int,
+        doc="Filter window size for highpass filter.",
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:137]: What are the units?  Pixels?

File: python/lsst/ip/isr/isrStatistics.py [hunk 3]
```diff
@@ -115,6 +117,43 @@ class IsrStatisticsTaskConfig(pexConfig.Config):
...
     )
 
+    doBiasShiftStatistics = pexConfig.Field(
+        dtype=bool,
+        doc="Measure number of image shifts in overscan?",
+        default=False,
+    )
+    biasShiftFilterOrder = pexConfig.Field(
+        dtype=int,
+        doc="Filter order for highpass filter.",
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:127]: Is this a polynomial order or something else?

File: python/lsst/ip/isr/isrStatistics.py [hunk 4]
```diff
@@ -215,11 +254,21 @@ def run(self, inputExp, ptc=None, overscanResults=None, **kwargs):
...
+
+        ampCorrelationResults = None
+        if self.config.doAmplifierCorrelationStatistics:
+            ampCorrelationResults = self.measureAmpCorrelations(inputExp, overscanResults)
+
         return pipeBase.Struct(
-            results={"CTI": ctiResults,
-                     "BANDING": bandingResults,
-                     "PROJECTION": projectionResults,
+            results={'CTI': ctiResults,
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:266]: I looks like one of these is still double quotation marks?

File: python/lsst/ip/isr/isrStatistics.py [hunk 5]
```diff
@@ -504,5 +554,171 @@ def copyCalibDistributionStatistics(self, inputExp, **kwargs):
...
+        outputStats = {}
+
+        detector = inputExp.getDetector()
+        for amp, overscans in zip(detector, overscanResults):
+            ampStats = {}
+            # Add fit back to data
+            rawOverscan = overscans.overscanImage.image.array + overscans.overscanFit
+
+            # Collapse array, skipping first three columns
+            rawOverscan = np.mean(rawOverscan[:, 3:], axis=1)
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:601]: Should this also be configurable?

File: python/lsst/ip/isr/isrStatistics.py [hunk 6]
```diff
@@ -504,5 +554,171 @@ def copyCalibDistributionStatistics(self, inputExp, **kwargs):
...
+        for region in shift_regions:
+            region_peak = np.argmax(shift_check[region[0]:region[1]]) + region[0]
+            if self._satisfies_flatness(region_peak, convolved[region_peak], overscanData):
+                shift_peaks.append(
+                    [float(convolved[region_peak]), float(region_peak),
+                     int(region[0]), int(region[1])])
+        return noise, shift_peaks
+
+    def _satisfies_flatness(self, shiftRow, shiftPeak, overscanData):
+        """
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:642]: docstring?

File: python/lsst/ip/isr/isrStatistics.py [hunk 7]
```diff
@@ -504,5 +554,171 @@ def copyCalibDistributionStatistics(self, inputExp, **kwargs):
...
+            # Scan for shifts
+            noise, shift_peaks = self._scan_for_shifts(rawOverscan)
+            ampStats['LOCAL_NOISE'] = float(noise)
+            ampStats['BIAS_SHIFTS'] = shift_peaks
+
+            outputStats[amp.getName()] = ampStats
+        return outputStats
+
+    def _scan_for_shifts(self, overscanData):
+        """
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:612]: docstring?

File: python/lsst/ip/isr/isrStatistics.py [hunk 8]
```diff
@@ -504,5 +554,171 @@ def copyCalibDistributionStatistics(self, inputExp, **kwargs):
...
+            preTrend = (2*preFit[0]*len(prerange) < shiftPeak)
+            postTrend = (2*postFit[0]*len(postrange) < shiftPeak)
+        else:
+            preTrend = (2*preFit[0]*len(prerange) > shiftPeak)
+            postTrend = (2*postFit[0]*len(postrange) > shiftPeak)
+
+        return (preTrend and postTrend)
+
+    def measureAmpCorrelations(self, inputExp, overscanResults):
+        """Measure number of bias shifts from overscan data.
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:660]: That doesn't look like the right docstring.

File: python/lsst/ip/isr/isrStatistics.py [hunk 9]
```diff
@@ -504,5 +554,171 @@ def copyCalibDistributionStatistics(self, inputExp, **kwargs):
...
+                (scalar or `lsst.afw.image.Image`).
+            ``overscanImage``
+                Image of the overscan region with the overscan
+                correction applied (`lsst.afw.image.Image`). This
+                quantity is used to estimate the amplifier read noise
+                empirically.
+
+        Returns
+        -------
+        outputStats : `dict` [`str`, [`dict` [`str`,`float]]
```
Comment (erykoff) [inline python/lsst/ip/isr/isrStatistics.py:683]: That is a complicated output!  Also missing a backtick.

**Review by erykoff — APPROVED** [review]
Some documentation fixes needed, nothing major.  I'm trusting that the algorithms are sound...

File: python/lsst/ip/isr/isrStatistics.py (hunk 3 above)
Comment (czwa) [reply python/lsst/ip/isr/isrStatistics.py:127]: I've clarified that this is a Butterworth highpass filter, which constrains the definition of order somewhat.

File: python/lsst/ip/isr/isrStatistics.py (hunk 4 above)
Comment (czwa) [reply python/lsst/ip/isr/isrStatistics.py:266]: I've fixed all instances of single quotes in this file.

File: python/lsst/ip/isr/isrStatistics.py (hunk 5 above)
Comment (czwa) [reply python/lsst/ip/isr/isrStatistics.py:601]: This is now `biasShiftColumnSkip`, and the already existing config renamed `biasShiftRowSkip`.

File: python/lsst/ip/isr/isrStatistics.py (hunk 9 above)
Comment (czwa) [reply python/lsst/ip/isr/isrStatistics.py:683]: Fixed.

File: python/lsst/ip/isr/isrStatistics.py (hunk 1 above)
Comment (czwa) [reply python/lsst/ip/isr/isrStatistics.py:142]: It's a signal/noise threshold.  I've updated the doc to indicate that.

**Pull request merged**